lib/_http_outgoing.js

```diff
@@ -526,7 +527,14 @@ OutgoingMessage.prototype.end = function(data, encoding, callback) {
 
   var self = this;
   function finish() {
-    self.emit('finish');
+    // NOTE: It is very important to wait for `prefinish` event.
+    // It is emitted only when the response becomes active in the pipeline.
+    // The ordering of requests/responses in `_http_server.js` is dependent
+    // on this, doing it in different order will result in crash.
+    if (self._prefinish)
+      self.emit('finish');
+    else
+      self.on('prefinish', finish);
```

Comment:
I feel like if this is possible then there's something else wrong with the logic. From the documentation:

> After [the finish] event, no more events will be emitted on the response object.

So how could this function ever be called more than once?